modules/xfeatures2d/src/tbmr.cpp

```diff
@@ -0,0 +1,646 @@
+                        imaAttribute[p][5] +=
+                            imaAttribute[root[r]][5]; // sum_yy
+
+                        if (rank[x] < rank[r])
+                        {
+                            // we merge p to r
+                            zpar[x] = r;
+                            root[r] = p;
+                            x = r;
+                        }
+                        else if (rank[r] < rank[p])
+                        {
+                            // merge r to p
+                            zpar[r] = p;
+                        }
+                        else
+                        {
+                            // same height
+                            zpar[r] = p;
+                            rank[p] += 1;
+                        }
+                    }
+                }
+            }
+        }
+
+        free(zpar);
+        free(root);
+        free(rank);
+        free(dejaVu);
+    }
+
+    void calculateTBMRs(const Mat &image, std::vector<Elliptic_KeyPoint> &tbmrs,
+                        const Mat &mask, float scale, int octave)
+    {
+        uint imSize = image.cols * image.rows;
+        uint maxArea =
+            static_cast<uint>(params.maxAreaRelative * imSize * scale);
+        uint minArea = static_cast<uint>(params.minArea * scale);
+
+        if (parent.empty() || parent.size != image.size)
+            parent = Mat(image.rows, image.cols, CV_32S);
+
+        if (imaAttributes.empty() || imaAttributes.size != image.size)
+            imaAttributes = Mat(image.rows, image.cols, CV_32SC(6));
+
+        calcMinMaxTree(image);
+
+        const Vec<uint, 6> *imaAttribute =
+            imaAttributes.ptr<const Vec<uint, 6>>();
+        const uint8_t *ima_ptr = image.ptr<const uint8_t>();
+        const uint *S_ptr = S.ptr<const uint>();
+        uint *parent_ptr = parent.ptr<uint>();
+
+        // canonization
+        for (uint i = 0; i < imSize; ++i)
+        {
+            uint p = S_ptr[i];
+            uint q = parent_ptr[p];
+            if (ima_ptr[parent_ptr[q]] == ima_ptr[q])
+                parent_ptr[p] = parent_ptr[q];
+        }
+
+        // TBMRs extraction
+        //------------------------------------------------------------------------
+        // small variant of the given algorithm in the paper. For each
+        // critical node having more than one child, we check if the
+        // largest region containing this node without any change of
+        // topology is above its parent, if not, discard this critical
+        // node.
+        //
+        // note also that we do not select the critical nodes themselves
+        // as final TBMRs
+        //--------------------------------------------------------------------------
+
+        uint *numSons = (uint *)calloc(imSize, sizeof(uint));
+        uint vecNodesSize = imaAttribute[S_ptr[0]][0];               // area
+        uint *vecNodes = (uint *)calloc(vecNodesSize, sizeof(uint)); // area
```

Comment:
> malloc/calloc/free/new/delete

Avoid using of manual memory management (it is not exception safe at least).
Consider using `cv::AutoBuffer` or `std::vector` instead.

Reply:
alright, done.